### PR TITLE
content clean cmd

### PR DESCRIPTION
**Pull request by dominik-przybyl-wttech**
#16

### CODE REVIEW

File: pkg/cfg/values.go [hunk 1]
```diff
@@ -140,4 +140,21 @@ type ConfigValues struct {
 			KeyBundleSymbolicName string `mapstructure:"key_bundle_symbolic_name" yaml:"key_bundle_symbolic_name"`
```
Comment (krystian-panek-wttech) [inline pkg/cfg/values.go:140]: There is no more values.go. Please merge main branch

File: pkg/content_clean.go [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+)
+
+func prepare(root string, config *cfg.ConfigValues) error {
+	if config.Content.ParentsBackupEnabled {
+		return doParentsBackup(root, config)
+	}
+	return nil
+}
+
+func BeforeClean(root string, config *cfg.ConfigValues) error {
```
Comment (krystian-panek-wttech) [inline pkg/content_clean.go:32]: basically I am trying to organise code into struct/services... you just implemented set of methods. this style is commonly used in Go community as well but I prefer to have style aligned everywhere.

how about 

```go
package content

type Cleaner struct {

}

func (c Cleaner) Clean(root string) error{

}
```

File: cmd/aem/package.go [hunk 3]
```diff
@@ -431,6 +431,13 @@ func pkgDefineDownloadFlags(cmd *cobra.Command) {
 	cmd.MarkFlagsMutuallyExclusive("pid", "path")
 }
 
+func pkgDefineUpdateFlags(cmd *cobra.Command) {
+	cmd.Flags().String("pid", "", "ID (group:name:version)'")
+	cmd.Flags().String("path", "", "Remote path on AEM repository")
```
Comment (krystian-panek-wttech) [inline cmd/aem/package.go:436]: all commands are `aem` scope so I am usually avoiding `aem` word in desc

File: pkg/common/httpx/httpx.go [hunk 4]
```diff
@@ -59,7 +60,7 @@ func DownloadWithOpts(opts DownloadOpts) error {
 }
 
 func DownloadWithChanged(opts DownloadOpts) (bool, error) {
-	if pathx.Exists(opts.File) {
+	if pathx.Exists(opts.File) && !opts.Override {
```
Comment (krystian-panek-wttech) [inline pkg/common/httpx/httpx.go:63]: override flag will force to always download so that whole method `DownloadWithChanged` does not make sense

File: pkg/content/content_clean.go [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+func (c Cleaner) BeforeClean(root string) error {
+	if c.config.ParentsBackupEnabled {
+		return c.doRootBackup(root)
+	}
+	return nil
+}
+
+func (c Cleaner) Clean(root string) error {
+	err := c.flattenFiles(root)
+	if err == nil {
```
Comment (krystian-panek-wttech) [inline pkg/content/content_clean.go:49]: in go early return is more preferred; please refactor this

```
if err := xxx(); err != nil {
 return err;
}
```

this will eliminate nesting code

File: pkg/content/content_clean.go [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+		err = deleteEmptyDirs(root)
+	}
+	return err
+}
+
+func eachFilesInDir(root string, processFileFunc func(path string) error) error {
+	infos, err := os.ReadDir(root)
+	for i := 0; i < len(infos) && err == nil; i++ {
+		if !infos[i].IsDir() {
+			err = processFileFunc(filepath.ToSlash(filepath.Join(root, infos[i].Name())))
```
Comment (krystian-panek-wttech) [inline pkg/content/content_clean.go:75]: this is wrong; early return; each err in iteration in your code could be overridden by another one which does not make sense

File: pkg/content/content_clean.go [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+		}
+	}
+	return err
+}
+
+func eachFiles(root string, processFileFunc func(string) error) error {
+	return filepath.Walk(root, func(path string, info fs.FileInfo, err error) error {
+		if err == nil && !info.IsDir() {
```
Comment (krystian-panek-wttech) [inline pkg/content/content_clean.go:83]: check for error - if not nil then return it; keep following this idiom as it is popular in go

File: pkg/common/httpx/httpx.go [hunk 8]
```diff
@@ -14,8 +14,9 @@ func FileNameFromURL(url string) string {
 }
 
 type DownloadOpts struct {
-	Url  string
-	File string
+	Url      string
```
Comment (krystian-panek-wttech) [inline pkg/common/httpx/httpx.go:17]: in Go abbr should be uppercase

File: pkg/content/content_clean.go [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+	err := c.flattenFiles(root)
+	if err == nil {
+		if c.config.ParentsBackupEnabled {
+			err = c.undoParentsBackup(root)
+		} else {
+			err = c.cleanParents(root)
+		}
+	}
+	if err == nil {
+		err = c.cleanDotContents(root)
```
Comment (krystian-panek-wttech) [inline pkg/content/content_clean.go:57]: if some of the following `if` will produce an error then further processing should be interrupted (`return` should be used here)

File: pkg/project/app_classic/aem/default/etc/aem.yml [hunk 10]
```diff
@@ -209,3 +209,53 @@ output:
     file: aem/home/var/log/aem.log
     # Controls where outputs and logs should be written to when format is 'text' (console|file|both)
     mode: console
+
+content:
+  files_dot_content:
```
Comment (krystian-panek-wttech) [inline pkg/project/app_classic/aem/default/etc/aem.yml:214]: how about

```yaml
files:
   deleted:
   flattened:
   // ...
```

File: pkg/project/app_classic/aem/default/etc/aem.yml [hunk 11]
```diff
@@ -209,3 +209,53 @@ output:
...
+    - patterns:
+        - "**/.vlt"
+        - "**/.vlt*.tmp"
+        - "**/install/*.jar"
+  files_flattened:
+    - "**/_cq_design_dialog/.content.xml"
+    - "**/_cq_dialog/.content.xml"
+    - "**/_cq_htmlTag/.content.xml"
+    - "**/_cq_template/.content.xml"
+  properties_skipped:
```
Comment (krystian-panek-wttech) [inline pkg/project/app_classic/aem/default/etc/aem.yml:231]: how about

```
properties:
  skipped:
```

File: pkg/project/app_classic/aem/default/etc/aem.yml [hunk 12]
```diff
@@ -209,3 +209,53 @@ output:
...
+        - "dam:parentAssetID"
+        - "dam:relativePath"
+  mixin_types_skipped:
+    - patterns: "dam:Thumbnails"
+      included_paths: [ "**/content/dam/*" ]
+    - patterns:
+        - "cq:ReplicationStatus"
+        - "mix:versionable"
+  namespaces_skipped: true
+  parents_backup_enabled: true
```
Comment (krystian-panek-wttech) [inline pkg/project/app_classic/aem/default/etc/aem.yml:260]:
```yaml
parents_backup:
  enabled: true
  suffix: ".bak"
```

File: pkg/package.go [hunk 13]
```diff
@@ -161,3 +161,36 @@ func (p Package) MarshalText() string {
...
 	return fmt.Sprintf("package '%s'", p.PID.String())
 }
+
+func (p Package) Create() error {
+	state, err := p.State()
+	if err != nil {
+		return err
+	}
+	if state.Exists {
+		return fmt.Errorf("%s > package '%s' cannot be create as it exists", p.manager.instance.ID(), p.PID.String())
```
Comment (krystian-panek-wttech) [inline pkg/package.go:171]: Created

File: pkg/package.go [hunk 14]
```diff
@@ -161,3 +161,36 @@ func (p Package) MarshalText() string {
...
+	}
+	return p.manager.Create(state.PID)
+}
+
+func (p Package) Update(filter []Filter) error {
```
Comment (krystian-panek-wttech) [inline pkg/package.go:176]: This is more like UpdateFilters method. In go there is no method overloading so the names need to be different

File: pkg/package.go [hunk 15]
```diff
@@ -161,3 +161,36 @@ func (p Package) MarshalText() string {
...
+	return p.manager.Update(state.Data.Path, state.PID, filter)
+}
+
+func (p Package) Download(localFile string) error {
+	state, err := p.State()
+	if err != nil {
+		return err
+	}
+	if !state.Exists {
+		return fmt.Errorf("%s > package '%s' cannot be download as it does not exist", p.manager.instance.ID(), p.PID.String())
```
Comment (krystian-panek-wttech) [inline pkg/package.go:193]: Downloaded

File: pkg/package_manager.go [hunk 16]
```diff
@@ -117,6 +119,96 @@ func (pm *PackageManager) IsSnapshot(localPath string) bool {
...
+	if err = fmtx.UnmarshalJSON(response.RawBody(), &status); err != nil {
+		return fmt.Errorf("%s > cannot create package '%s'; cannot parse response: %w", pm.instance.ID(), pid, err)
+	}
+	if !status.Success {
+		return fmt.Errorf("%s > cannot create package '%s'; unexpected status: %s", pm.instance.ID(), pid, status.Message)
+	}
+	log.Infof("%s > create package '%s'", pm.instance.ID(), pid)
+	return nil
+}
+
```
Comment (krystian-panek-wttech) [inline pkg/package_manager.go:150]: Define these types under pkg subpackage

File: pkg/project/app_classic/aem/default/etc/aem.yml [hunk 17]
```diff
@@ -4,10 +4,10 @@ instance:
   # Full details of local or remote instances
   config:
     local_author:
-      active: [[.Env.AEM_AUTHOR_ACTIVE | default true ]]
-      http_url: [[.Env.AEM_AUTHOR_HTTP_URL | default "http://127.0.0.1:4502" ]]
-      user: [[.Env.AEM_AUTHOR_USER | default "admin" ]]
-      password: [[.Env.AEM_AUTHOR_PASSWORD | default "admin" ]]
+      active: [ [ .Env.AEM_AUTHOR_ACTIVE | default true ] ]
```
Comment (krystian-panek-wttech) [inline pkg/project/app_classic/aem/default/etc/aem.yml:7]: Please revert these changes. Unnecessarily touched

File: pkg/content_download.go [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+	config *content.Opts
+}
+
+func NewDownloader(config *content.Opts) *Downloader {
+	return &Downloader{
+		config: config,
+	}
+}
+
+//go:embed content/aemc_content
```
Comment (krystian-panek-wttech) [inline pkg/content_download.go:28]: How about 'vault' subpackage name instead of aemc_content?

File: pkg/content_download.go [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+	}
+	_, err = filex.ArchiveWithChanged("/tmp/aemc_content", "/tmp/aemc_content.zip")
+	if err != nil {
+		return err
+	}
+	_, err = packageManager.Upload("/tmp/aemc_content.zip")
+	if err != nil {
+		return err
+	}
+	err = packageManager.Build("/etc/packages/my_packages/aemc_content.zip")
```
Comment (krystian-panek-wttech) [inline pkg/content_download.go]: Constant path makes increases a chance in collisions when running pkg content commands twice on the same version. Maybe some randomised name need to be introduced here. Check how was that impl in GAP

File: pkg/content_move.go [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+	config *content.Opts
+}
+
+func NewMover(config *content.Opts) *Mover {
+	return &Mover{
+		config: config,
+	}
+}
+
+func (c Mover) Move(scrPackageManager *PackageManager, descPackageManager *PackageManager, filter string) error {
```
Comment (krystian-panek-wttech) [inline pkg/content_move.go]: How about cleaning redundant / tmp packages on source and target instances?

File: cmd/aem/content.go [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+	_ = cmd.MarkFlagRequired("filter-path")
+	cmd.Flags().Bool("clean", true, "Clean content after download")
+	return cmd
+}
+
+func (c *CLI) contentMoveCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "move",
+		Aliases: []string{"mv"},
+		Short:   "Move content from one instance to another",
```
Comment (krystian-panek-wttech) [inline cmd/aem/content.go:83]: it's more like `copy` ;) content is copied not moved

File: cmd/aem/content.go [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+				c.Error(fmt.Errorf("content move failed: %w", err))
+				return
+			}
+			c.Ok("content moved")
+		},
+	}
+	cmd.Flags().String("src-instance-url", "", "Source instance URL")
+	cmd.Flags().String("src-instance-id", "", "Source instance ID")
+	cmd.MarkFlagsMutuallyExclusive("src-instance-url", "src-instance-id")
+	cmd.Flags().String("desc-instance-url", "", "Destination instance URL")
```
Comment (krystian-panek-wttech) [inline cmd/aem/content.go:108]: `dest` not `desc`

File: cmd/aem/content.go [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+	cmd.MarkFlagsMutuallyExclusive("desc-instance-url", "desc-instance-id")
+	cmd.Flags().String("filter-path", "", "Filter path")
+	_ = cmd.MarkFlagRequired("filter-path")
+	cmd.Flags().Bool("clean", true, "Clean content before move")
+	return cmd
+}
+
+func (c *CLI) determineInstance(cmd *cobra.Command, urlParamName string, idParamName string, errorMsg string) (*pkg.Instance, error) {
+	var instance *pkg.Instance
+	if cmd.Flags().Lookup(urlParamName) != nil {
```
Comment (krystian-panek-wttech) [inline cmd/aem/content.go:119]: why not to assume that empty string also means undefined value

File: cmd/aem/content.go [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+			instance = c.aem.InstanceManager().NewByID(id)
+		}
+	}
+	if instance == nil {
+		return nil, fmt.Errorf(errorMsg)
+	}
+	return instance, nil
+}
+
+func (c *CLI) determineRootPath(cmd *cobra.Command) (string, error) {
```
Comment (krystian-panek-wttech) [inline cmd/aem/content.go:138]: receiver scope is global but the method name is quite generic

File: cmd/aem/package.go [hunk 25]
```diff
@@ -430,6 +433,24 @@ func pkgDefineFlags(cmd *cobra.Command) {
...
+	cmd.Flags().String("pid", "", "ID (group:name:version)'")
+	cmd.Flags().String("file", "", "Local path on file system")
+	cmd.Flags().String("path", "", "Remote path on AEM repository")
+	cmd.MarkFlagsMutuallyExclusive("pid", "path")
+}
+
+func pkgDefineUpdateFlags(cmd *cobra.Command) {
+	cmd.Flags().String("pid", "", "ID (group:name:version)'")
+	cmd.Flags().String("path", "", "Remote path on AEM repository")
+	cmd.Flags().StringSlice("filter-root", nil, "Filter root path(s) on AEM repository")
```
Comment (krystian-panek-wttech) [inline cmd/aem/package.go:446]: prefer empty list than nil ...

File: examples/docker/src/aem/default/etc/aem.yml [hunk 26]
```diff
@@ -208,3 +208,47 @@ output:
     file: aem/home/var/log/aem.log
     # Controls where outputs and logs should be written to when format is 'text' (console|file|both)
     mode: console
+
+content:
+  files_deleted:
+    - patterns: "**/_jcr_content/folderThumbnail*"
```
Comment (krystian-panek-wttech) [inline examples/docker/src/aem/default/etc/aem.yml:214]: I am not sure if dam thumbs should be ottb deleted

File: examples/docker/src/aem/default/etc/aem.yml [hunk 27]
```diff
@@ -208,3 +208,47 @@ output:
...
+        - "dam:assetState"
+        - "dc:modified"
+        - "*_x0040_*"
+        - "cq:name"
+        - "cq:parentPath"
+        - "dam:copiedAt"
+        - "dam:parentAssetID"
+        - "dam:relativePath"
+  mixin_types_skipped:
+    - patterns: "dam:Thumbnails"
```
Comment (krystian-panek-wttech) [inline examples/docker/src/aem/default/etc/aem.yml:248]: this as well

File: pkg/cfg/defaults.go [hunk 28]
```diff
@@ -92,4 +92,71 @@ func (c *Config) setDefaults() {
...
+		},
+		map[string]interface{}{
+			"patterns": []string{
+				"cq:ReplicationStatus",
+				"mix:versionable",
+			},
+		},
+	})
+	v.SetDefault("content.namespaces_skipped", true)
+	v.SetDefault("content.parents_backup_enabled", true)
```
Comment (krystian-panek-wttech) [inline pkg/cfg/defaults.go:161]: you were removing that option, weren't you?

File: pkg/common/httpx/httpx.go [hunk 29]
```diff
@@ -51,25 +53,25 @@ func DownloadWithOpts(opts DownloadOpts) error {
 	if err := pathx.Ensure(filepath.Dir(fileTmp)); err != nil {
 		return err
 	}
-	res, err := http.Get(opts.URL)
+	res, err := client.R().Get(opts.URL)
```
Comment (krystian-panek-wttech) [inline pkg/common/httpx/httpx.go:56]: why?

File: pkg/common/httpx/httpx.go [hunk 30]
```diff
@@ -51,25 +53,25 @@ func DownloadWithOpts(opts DownloadOpts) error {
...
 	if err != nil {
 		return fmt.Errorf("cannot download from URL '%s' to file '%s': %w", opts.URL, opts.File, err)
 	}
-	defer res.Body.Close()
-	if res.StatusCode != http.StatusOK {
-		return fmt.Errorf("cannot download from URL '%s' to file '%s': %s", opts.URL, opts.File, res.Status)
+	defer res.RawBody().Close()
```
Comment (krystian-panek-wttech) [inline pkg/common/httpx/httpx.go:60]: why?

File: pkg/content/aemc_content/META-INF/MANIFEST.MF [hunk 31]
```diff
@@ -0,0 +1,4 @@
+Manifest-Version: 1.0
+Content-Package-Id: my_packages:aemc_content
```
Comment (krystian-panek-wttech) [inline pkg/content/aemc_content/META-INF/MANIFEST.MF:2]: `aemc:content:{{timestamp}}` ?

File: pkg/content/aemc_content/META-INF/vault/properties.xml [hunk 32]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE properties SYSTEM "http://java.sun.com/dtd/properties.dtd">
+<properties>
+    <entry key="packageType">mixed</entry>
+    <entry key="packageFormatVersion">2</entry>
+    <entry key="group">my_packages</entry>
```
Comment (krystian-panek-wttech) [inline pkg/content/aemc_content/META-INF/vault/properties.xml:6]: aemc
content
{{timestamp}} ?

File: pkg/content/content_clean.go (hunk 5 above)
Comment (krystian-panek-wttech) [reply pkg/content/content_clean.go:51]: still to be fixed @dominik-przybyl-wttech

File: pkg/content_move.go [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+func NewMover(config *content.Opts) *Mover {
+	return &Mover{
+		config: config,
+	}
+}
+
+func (c Mover) Move(scrPackageManager *PackageManager, descPackageManager *PackageManager, filter string, clean bool) error {
+	err := NewDownloader(c.config).Download(scrPackageManager, "/tmp/aemc_content", filter, clean)
```
Comment (krystian-panek-wttech) [inline pkg/content_move.go:19]: don't use hardcoded `/tmp` ; it will not work on Windows

in AEMC there is dedicated tmp dir defined in base opts

File: pkg/package.go [hunk 34]
```diff
@@ -161,3 +161,36 @@ func (p Package) MarshalText() string {
...
+	return p.manager.Create(state.PID)
+}
+
+func (p Package) UpdateFilters(filters []Filter) error {
+	state, err := p.State()
+	if err != nil {
+		return err
+	}
+	if !state.Exists {
+		return fmt.Errorf("%s > filters for package '%s' cannot be updated as it does not exist", p.manager.instance.ID(), p.PID.String())
```
Comment (krystian-panek-wttech) [inline pkg/package.go:182]: package '%s' does not exist so the filters cannot be updated

File: pkg/package_manager.go [hunk 35]
```diff
@@ -120,6 +122,96 @@ func (pm *PackageManager) IsSnapshot(localPath string) bool {
...
+type Filter struct {
+	Root  string `json:"root"`
+	Rules []Rule `json:"rules"`
+}
+
+type Rule struct {
+	Modifier string `json:"modifier"`
```
Comment (krystian-panek-wttech) [inline pkg/package_manager.go:160]: modifier is a mode? merge|update etc?

File: pkg/project/app_classic/aem/default/etc/aem.yml [hunk 36]
```diff
@@ -209,3 +209,47 @@ output:
...
+        - "dam:parentAssetID"
+        - "dam:relativePath"
+  mixin_types_skipped:
+    - patterns: "dam:Thumbnails"
+      included_paths: [ "**/content/dam/*" ]
+    - patterns:
+        - "cq:ReplicationStatus"
+        - "mix:versionable"
+  namespaces_skipped: true
+  parents_backup_enabled: true
```
Comment (krystian-panek-wttech) [inline pkg/project/app_classic/aem/default/etc/aem.yml:255]: you were removing parents backup as far as I know

File: pkg/project/instance/aem/default/etc/aem.yml [hunk 37]
```diff
@@ -208,3 +208,47 @@ output:
     file: aem/home/var/log/aem.log
     # Controls where outputs and logs should be written to when format is 'text' (console|file|both)
     mode: console
+
+content:
+  files_deleted:
+    - patterns: "**/_jcr_content/folderThumbnail*"
```
Comment (krystian-panek-wttech) [inline pkg/project/instance/aem/default/etc/aem.yml:214]: do not remove thumbs

File: pkg/project/instance/aem/default/etc/aem.yml [hunk 38]
```diff
@@ -208,3 +208,47 @@ output:
...
+        - "dam:assetState"
+        - "dc:modified"
+        - "*_x0040_*"
+        - "cq:name"
+        - "cq:parentPath"
+        - "dam:copiedAt"
+        - "dam:parentAssetID"
+        - "dam:relativePath"
+  mixin_types_skipped:
+    - patterns: "dam:Thumbnails"
```
Comment (krystian-panek-wttech) [inline pkg/project/instance/aem/default/etc/aem.yml:248]: do not remove thumbs support OOTB

File: pkg/package.go [hunk 39]
```diff
@@ -161,3 +161,36 @@ func (p Package) MarshalText() string {
...
+	if err != nil {
+		return err
+	}
+	if state.Exists {
+		return fmt.Errorf("%s > package '%s' cannot be created as it already exists", p.manager.instance.ID(), p.PID.String())
+	}
+	return p.manager.Create(state.PID)
+}
+
+func (p Package) UpdateFilters(filters []Filter) error {
```
Comment (krystian-panek-wttech) [inline pkg/package.go:176]: type `Filter` is too generic for package `pkg`

File: pkg/package_manager.go (hunk 16 above)
Comment (krystian-panek-wttech) [reply pkg/package_manager.go:153]: `pkg.pkg.Filter` or `pkg.vault.Filter`

File: pkg/content_download.go [hunk 40]
```diff
@@ -0,0 +1,84 @@
...
+	config *content.Opts
+}
+
+func NewDownloader(config *content.Opts) *Downloader {
+	return &Downloader{
+		config: config,
+	}
+}
+
+func (c Downloader) DownloadPackage(packageManager *PackageManager, pid string, roots []string, filter string) (string, error) {
```
Comment (krystian-panek-wttech) [inline pkg/content_download.go:26]: I am seeing it more like 

`PackageManager.DownloadPaths(localFile, pid, rootPaths, filter)`

as it has so close relationship with pkg manager; maybe it just needs to be a part of pkg mgr?:) I am open for it

File: pkg/content_download.go [hunk 41]
```diff
@@ -0,0 +1,84 @@
...
+		return "", err
+	}
+	tmpResultFile := filepath.Join(c.config.BaseOpts.TmpDir, filepath.Base(remotePath))
+	if err := packageManager.Download(remotePath, tmpResultFile); err != nil {
+		return "", err
+	}
+	return tmpResultFile, nil
+}
+
+func (c Downloader) DownloadContent(packageManager *PackageManager, pid string, root string, roots []string, filter string, clean bool, unpack bool) error {
```
Comment (krystian-panek-wttech) [inline pkg/content_download.go:47]: this method should be in another place, rather unpacking should be disconnected from downloading

File: pkg/facade.go [hunk 42]
```diff
@@ -32,6 +34,7 @@ func NewAEM(config *cfg.Config) *AEM {
 	result.project = project.New(result.config)
 	result.baseOpts = base.NewOpts(result.config)
 	result.javaOpts = java.NewOpts(result.baseOpts)
+	result.contentOpts = content.NewOpts(result.baseOpts)
```
Comment (krystian-panek-wttech) [inline pkg/facade.go:37]: `result.contentManager = content.newManager()`

File: pkg/package.go [hunk 43]
```diff
@@ -161,3 +161,36 @@ func (p Package) MarshalText() string {
 func (p Package) String() string {
 	return fmt.Sprintf("package '%s'", p.PID.String())
 }
+
+func (p Package) Create(rootPaths []string, filterFile string) (string, error) {
```
Comment (krystian-panek-wttech) [inline pkg/package.go:165]: keep consistent naming, previously `filter` is here named as `filterFile` right?

File: pkg/package_manager.go [hunk 44]
```diff
@@ -35,6 +43,7 @@ func NewPackageManager(res *Instance) *PackageManager {
 
 	return &PackageManager{
 		instance: res,
+		config:   res.manager.aem.baseOpts,
```
Comment (krystian-panek-wttech) [inline pkg/package_manager.go:46]: this field is not needed; in a place you are accessing it use `instance` field directly, or introduce shorthand private method

File: pkg/package_manager.go [hunk 45]
```diff
@@ -130,6 +139,147 @@ func (pm *PackageManager) IsSnapshot(localPath string) bool {
 	return stringsx.MatchSome(pathx.Normalize(localPath), pm.SnapshotPatterns)
 }
 
+//go:embed package/default
+var defaultPackage embed.FS
```
Comment (krystian-panek-wttech) [inline pkg/package_manager.go:143]: I prefer naming `topic_trait` more than `trait_topic` so... `packageDefault`

**Review by krystian-panek-wttech — CHANGES_REQUESTED** [review]

File: cmd/aem/content.go [hunk 46]
```diff
@@ -0,0 +1,187 @@
...
+			c.SetOutput("dir", dir)
+			c.Ok("content synchronized")
+		},
+	}
+	cmd.Flags().StringP("dir", "d", "", "JCR root path")
+	_ = cmd.MarkFlagRequired("dir")
+	cmd.Flags().StringSliceP("filter-roots", "r", nil, "Vault filter root paths")
+	cmd.Flags().StringP("filter-file", "f", "", "Vault filter file path")
+	cmd.MarkFlagsMutuallyExclusive("filter-file", "filter-roots")
+	cmd.Flags().Bool("clean", true, "Normalize content after downloading")
```
Comment (krystian-panek-wttech) [inline cmd/aem/content.go:118]: BoolP "-c"

File: cmd/aem/content.go [hunk 47]
```diff
@@ -0,0 +1,187 @@
...
+			c.Ok("content copied")
+		},
+	}
+	cmd.Flags().StringP("instance-target-url", "u", "", "Destination instance URL")
+	cmd.Flags().StringP("instance-target-id", "i", "", "Destination instance ID")
+	cmd.MarkFlagsMutuallyExclusive("instance-target-url", "instance-target-id")
+	cmd.Flags().StringSliceP("filter-roots", "r", []string{}, "Vault filter root paths")
+	cmd.Flags().StringP("filter-file", "f", "", "Vault filter file path")
+	cmd.MarkFlagsMutuallyExclusive("filter-file", "filter-roots")
+	cmd.Flags().Bool("clean", false, "Normalize content while copying")
```
Comment (krystian-panek-wttech) [inline cmd/aem/content.go:161]: BoolP

File: examples/docker/src/aem/default/etc/aem.yml [hunk 48]
```diff
@@ -220,3 +220,43 @@ output:
     file: aem/home/var/log/aem.log
     # Controls where outputs and logs should be written to when format is 'text' (console|file|both)
     mode: console
+
+content:
+  files_deleted:
```
Comment (krystian-panek-wttech) [inline examples/docker/src/aem/default/etc/aem.yml:225]: probably `clean` scope to be added here

File: pkg/common/filex/archive.go [hunk 49]
```diff
@@ -57,10 +58,17 @@ func Unarchive(sourceFile string, targetDir string) error {
 	if !pathx.Exists(sourceFile) {
 		return fmt.Errorf("cannot unarchive file '%s' to dir '%s' as source file does not exist", sourceFile, targetDir)
 	}
-	if err := pathx.Ensure(targetDir); err != nil {
-		return err
+	//if err := pathx.Ensure(targetDir); err != nil {
```
[the remark this answers is not in the document]
Comment (krystian-panek-wttech) [reply pkg/common/filex/archive.go:61]: bump @dominik-przybyl-wttech  ETA?

File: pkg/content_manager.go [hunk 50]
```diff
@@ -0,0 +1,109 @@
...
+	return cm.instance.PackageManager()
+}
+
+func (cm *ContentManager) tmpDir() string {
+	return cm.instance.manager.aem.baseOpts.TmpDir
+}
+
+func (cm *ContentManager) Download(localFile string, opts PackageCreateOpts) error {
+	if opts.PID == "" {
+		opts.PID = fmt.Sprintf("aemc:content-download:%s-SNAPSHOT", time.Now().Format("2006.102.304"))
```
Comment (krystian-panek-wttech) [inline pkg/content_manager.go:31]: nice pkg thumbnail need to be added

File: pkg/package.go [hunk 51]
```diff
@@ -161,3 +161,37 @@ func (p Package) MarshalText() string {
 func (p Package) String() string {
 	return fmt.Sprintf("package '%s'", p.PID.String())
 }
+
+func (p Package) Create(opts PackageCreateOpts) (string, error) {
```
Comment (krystian-panek-wttech) [inline pkg/package.go:165]: `CreateWithChanged` ?

File: pkg/package.go [hunk 52]
```diff
@@ -161,3 +161,37 @@ func (p Package) MarshalText() string {
...
+		return "", err
+	}
+	if state.Exists {
+		return "", fmt.Errorf("%s > package '%s' cannot be created as it already exists", p.manager.instance.ID(), p.PID.String())
+	}
+	opts.PID = state.PID
+	return p.manager.Create(opts)
+}
+
+func (p Package) UpdateFilters(filters []PackageFilter) error {
```
Comment (krystian-panek-wttech) [inline pkg/package.go:177]: `UpdateFiltersWithChanged` ?

File: pkg/cfg/defaults.go (hunk 28 above)
Comment (dominik-przybyl-wttech) [reply pkg/cfg/defaults.go:167]: I removed option: parents_backup_suffix, now is .bak as default

File: pkg/content_manager.go (hunk 50 above)
Comment (dominik-przybyl-wttech) [reply pkg/content_manager.go:31]: done

File: pkg/common/httpx/httpx.go (hunk 29 above)
Comment (dominik-przybyl-wttech) [reply pkg/common/httpx/httpx.go:56]: bugfix

File: pkg/common/httpx/httpx.go (hunk 30 above)
Comment (dominik-przybyl-wttech) [reply pkg/common/httpx/httpx.go:60]: bugfix

File: pkg/package.go [hunk 53]
```diff
@@ -161,3 +176,62 @@ func (p Package) MarshalText() string {
...
+	}
+	if !state.Exists {
+		opts.PID = state.PID
+		_, err = p.manager.Create(opts)
+		return true, err
+	}
+	return false, nil
+}
+
+func (p Package) UpdateFiltersWithChanged(filters []PackageFilter) error {
```
Comment (krystian-panek-wttech) [inline pkg/package.go:203]: where is boolean => for changed value?

don't pretend that it is implemented.. :D

Comment (krystian-panek-wttech) [reply pkg/package.go:203]: I mean methods with `WithChanged` should always return `bool`

File: cmd/aem/content.go [hunk 54]
```diff
@@ -0,0 +1,196 @@
...
+			dir, err := determineContentDir(cmd)
+			if err != nil {
+				c.Error(err)
+				return
+			}
+			clean, _ := cmd.Flags().GetBool("clean")
+			filterRoots, _ := cmd.Flags().GetStringSlice("filter-roots")
+			filterFile, _ := cmd.Flags().GetString("filter-file")
+			if len(filterRoots) == 0 && filterFile == "" {
+				filterRoots = append(filterRoots, strings.Split(dir, content.JCRRoot)[1])
```
Comment (krystian-panek-wttech) [inline cmd/aem/content.go:113]: What if dir will have trailing slash? 🙂

File: cmd/aem/package.go [hunk 55]
```diff
@@ -431,6 +445,41 @@ func pkgDefineFlags(cmd *cobra.Command) {
 	cmd.MarkFlagsMutuallyExclusive("pid", "file", "path")
 }
 
+func pkgDefineDownloadFlags(cmd *cobra.Command) {
+	cmd.Flags().String("pid", "", "ID (group:name:version)'")
+	cmd.Flags().String("path", "", "Remote path on AEM repository")
```
Comment (krystian-panek-wttech) [inline cmd/aem/package.go:450]: Remote repository path

File: pkg/package_manager.go [hunk 56]
```diff
@@ -130,6 +134,150 @@ func (pm *PackageManager) IsSnapshot(localPath string) bool {
...
+	return filters
+}
+
+type PackageFilterRule struct {
+	Modifier string `json:"modifier"`
+	Pattern  string `json:"pattern"`
+}
+
+func (pm *PackageManager) UpdateFilters(remotePath string, pid string, filters []PackageFilter) error {
+	log.Infof("%s > updating package '%s'", pm.instance.ID(), pid)
```
Comment (krystian-panek-wttech) [inline pkg/package_manager.go:233]: updating filters of package

File: pkg/package_manager.go [hunk 57]
```diff
@@ -130,6 +134,150 @@ func (pm *PackageManager) IsSnapshot(localPath string) bool {
...
+		SetMultipartFormData(map[string]string{
+			"path":        remotePath,
+			"packageName": pidConfig.Name,
+			"groupName":   pidConfig.Group,
+			"version":     pidConfig.Version,
+			"filter":      string(filtersJson),
+		}).
+		Post(UpdatePath)
+	if err != nil {
+		return fmt.Errorf("%s > cannot update filters for package '%s': %w", pm.instance.ID(), pid, err)
```
Comment (krystian-panek-wttech) [inline pkg/package_manager.go:252]: of package - everywhere

File: pkg/package_manager.go [hunk 58]
```diff
@@ -251,9 +399,9 @@ func (pm *PackageManager) installHTML(remotePath string) error {
 		if err != nil {
 			return fmt.Errorf("%s > cannot install package '%s': cannot open HTML report file '%s'", pm.instance.ID(), remotePath, htmlFilePath)
 		}
-		defer htmlFile.Close()
+		defer func() { _ = htmlFile.Close() }()
```
Comment (krystian-panek-wttech) [inline pkg/package_manager.go:402]: Have you tested it? @dominik-przybyl-wttech

**Review by krystian-panek-wttech — APPROVED** [review]
